### PR TITLE
DB-11342 Fix ArrayIndexOutOfBoundsException with NLJ and triggers

### PR DESCRIPTION
**Pull request by msirek**
# Short Description
Fixes ArrayIndexOutOfBoundsException and NPE during execution of a triggers which accesses trigger referencing rows via nested loop join by making sure the correct TriggerExecutionContext is always being used.
# Long Description
- Gets rid of the problematic pushTriggerExecutionContextFromActivation.  The TriggerExecutionContext should come from the local LCC from the thread's context manager.
- Updates the trigger rowHolder to always point to the correct TriggerExecutionContext upon deserialization.
- In TriggerNewTransitionRows.getDataSet, use the TriggerExecutionContext from the class field, as it should always be correct.  The object is instantiated when the operation tree's init functions are called and allocates a new TriggerExecutionContext for the thread.

# How to test
The specific test from LiveWire was added to the ITs.

### CODE REVIEW

**Comment by msirek** [comment]
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes

**Comment by spliceci** [comment]
TEST SUCCEEDED +1

**Review by arnaud-splice — APPROVED** [review]
+1
Please wait for @OlegMazurov's review before merging this one

**Review by ascend1 — APPROVED** [review]

**Review by dgomezferro — APPROVED** [review]
I love the cleanup!

**Review by carolp-503 — APPROVED** [review]
+1

File: splice_machine/src/main/java/com/splicemachine/derby/stream/iterator/GetNLJoinIterator.java
Comment (OlegMazurov) [inline splice_machine/src/main/java/com/splicemachine/derby/stream/iterator/GetNLJoinIterator.java:83]: Creating a new ContextManager is unnecessary and undesirable here.
If somehow this code is still required, a proper fix needs to be implemented/extended at a higher level.

**Review by OlegMazurov — CHANGES_REQUESTED** [review]
This is a good cleanup. I ask for it to be extended to remove unnecessary ContextManager creation and manipulation (see TriggerExecutionContext.pushLanguageConnectionContextToCM()).
Also, this PR should be postponed (or rather combined with) until after current trigger regressions are fixed (DB-11530, DB-11535).

**Comment by msirek** [comment]
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes

File: splice_machine/src/main/java/com/splicemachine/derby/stream/iterator/GetNLJoinIterator.java
Comment (msirek) [reply splice_machine/src/main/java/com/splicemachine/derby/stream/iterator/GetNLJoinIterator.java]: Benchmark results after latest code push:
java -jar SpliceBenchmark.jar -S 10000 -c 1 -n 10000
```
##### INSERT vs. UPDATE Trigger Benchmark #####
2021-02-25 21:11:27    Setup
2021-02-25 21:11:30    Create and populate tables
2021-02-25 21:11:45    Statistics:
	CREATE	calls: 20000	time: 15 s	avg: 1 ms
2021-02-25 21:11:45    Analyze schema, create triggers
2021-02-25 21:11:53    Do inserts
2021-02-25 21:12:53    Statistics:
	INSERT_T	calls: 8074	time: 29 s	avg: 4 ms
	INSERT_NOT	calls: 8073	time: 30 s	avg: 4 ms
2021-02-25 21:13:08    Statistics:
	INSERT_T	calls: 1926	time: 6 s	avg: 4 ms
	INSERT_NOT	calls: 1927	time: 7 s	avg: 4 ms
2021-02-25 21:13:08    Do updates
2021-02-25 21:14:08    Statistics:
	UPDATE_T	calls: 4708	time: 31 s	avg: 7 ms
	UPDATE_NOT	calls: 4707	time: 29 s	avg: 6 ms
2021-02-25 21:15:08    Statistics:
	UPDATE_T	calls: 4891	time: 31 s	avg: 6 ms
	UPDATE_NOT	calls: 4891	time: 28 s	avg: 6 ms
2021-02-25 21:15:13    Statistics:
	UPDATE_T	calls: 401	time: 2 s	avg: 6 ms
	UPDATE_NOT	calls: 402	time: 2 s	avg: 6 ms
2021-02-25 21:15:13    Finished
```

**Comment by spliceci** [comment]
TEST FAILED
http://nexus.splicemachine.com:8080/job/spliceengine-PR/job/Splice-PR-Branch/2540/

**Comment by msirek** [comment]
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes

**Comment by spliceci** [comment]
TEST SUCCEEDED +1

File: splice_machine/src/main/java/com/splicemachine/derby/stream/iterator/GetNLJoinIterator.java
Comment (OlegMazurov) [inline splice_machine/src/main/java/com/splicemachine/derby/stream/iterator/GetNLJoinIterator.java:71]: I still have reservations about this code. A new ContextManager is created underneath with a new LCC. I see that there are leftover ContextManagers stuck in ThreadLocal of Spark executor threads (not necessarily from this code). That's one step (if not there already) from a fully fledged memory leak.
The question remains, why is it needed? Should it be addressed at a higher level? Don't nested ContextManagers in SpliceTransactionManager help?

**Comment by OlegMazurov** [comment]
I see that the latest changes address my previous comment and DB-11535 (performance regression in INSERT triggers), though not DB-11530 (spurious write conflicts).
I don't object integrating this PR but strongly believe that more work is required on fixing trigger functionality and code improvement.

**Comment by carolp-503** [comment]
Thanks Oleg!  Yes, this is definitely not the last work we'll be doing in this area.

**Pull request merged**
